taichi/program/kernel.h

```diff
@@ -37,10 +38,18 @@ class Kernel : public Callable {
     void set_extra_arg_int(int i, int j, int32 d);
 
     void set_arg_external_array(int arg_id,
-                                uint64 ptr,
+                                uintptr_t ptr,
                                 uint64 size,
                                 bool is_device_allocation);
 
+    void set_arg_external_array_with_shape(int arg_id,
+                                           uintptr_t ptr,
+                                           uint64 size,
+                                           const std::vector<int64> &shape,
+                                           bool is_device_allocation);
+
+    void set_arg_external_ndarray(int arg_id, const Ndarray &arr);
```

Comment:
oh sorry one more nitpicking: let's name this `set_arg_ndarray` since there no internal/external concept ndarrays ;)

Reply:
Done :)